src/Analysis/Ast/Impl/Types/PythonUnionType.cs

```diff
@@ -28,15 +28,19 @@ internal sealed class PythonUnionType : LocatedMember, IPythonUnionType {
 
         public PythonUnionType(IEnumerable<IPythonType> types, IPythonModule declaringModule)
             : base(declaringModule.Interpreter.ModuleResolution.GetSpecializedModule("typing")) {
-            _types.UnionWith(types);
+            _types.UnionWith(types.Where(t => !this.Equals(t)));
```

Comment:
Do either of the constructors end up mattering? Or is it just `Add` that was the problem? I'm not sure how you could end up ever having an object "equal" to the current one if you're in the constructor for `this`, but I could be wrong.

Reply:
It basically gets confused on
```python
    def get_if_str(self, obj: Union[str, Factory]):
        """
        Returns object from the registry if ``obj`` type is string.
        """
        if type(obj) is str:
            return self.get(obj)
        return obj
```
when calculating return value. Another option is to introduce deep compare to `Union` so `!currentType.Equals(valueType)` would trigger. But it would be narrower fix as there may be other calls to `Add` ending up with circular unions.

```cs
            if (!_fromAnnotation && !currentType.Equals(valueType)) {
                var type = PythonUnionType.Combine(currentType, valueType);
                // Track instance vs type info.
                StaticReturnValue = value is IPythonInstance ? type.CreateInstance(ArgumentSet.WithoutContext) : (IMember)type;
            }

```

Reply:
Hm, okay.